File: dotnet/src/LabelZoom.Sdk/Conversion/ConversionSourceBuilder.cs
namespace LabelZoom.Sdk.Conversion
{
    /// <summary>
    /// Chooses the target format. One class covers all eleven targets.
    /// </summary>
    /// <remarks>
    /// There is no <c>ToUrl</c>: <c>URL</c> is a source-only fetch instruction, and
    /// <see cref="TargetFormat"/> has no member for it. Attempting one is a compile error rather
    /// than a runtime 404.
    /// </remarks>
    public sealed class ConversionSourceBuilder
    {
        private readonly LabelZoomClient _client;
        private readonly SourceFormat _source;
        private readonly byte[] _body;
        private readonly string _contentType;
        private readonly ConversionParameters _parameters;

        internal ConversionSourceBuilder(
            LabelZoomClient client, SourceFormat source, byte[] body, string contentType)
        {
            _client = client;
            _source = source;
            _body = body;
            _contentType = contentType;
            _parameters = new ConversionParameters();
        }

        /// <summary>
        /// How the <b>source's</b> absolute positions are interpreted, in dots per inch: dots for
        /// printer languages, pixels for bitmap images, an override of the document's dpi for
        /// LabelZoom XML/JSON. Not applicable to PDF sources (vector). This is the source-side dpi;
        /// use <see cref="ConversionTargetBuilder.WithDpi"/> to author the output at a resolution,
        /// and both may be set when the chosen format pair supports a dpi on each side.
        /// </summary>
        public ConversionSourceBuilder WithDpi(int dpi)
        {
            if (dpi <= 0)
            {
                throw new LabelZoomValidationException(nameof(dpi), "DPI must be greater than zero.");
            }

            _parameters.Set("sourceDpi", dpi);
            return this;
        }

        /// <summary>Selects the target format.</summary>
        public ConversionTargetBuilder To(TargetFormat target) =>
            new ConversionTargetBuilder(_client, _source, target, _body, _contentType, _parameters);

        /// <summary>Converts to ZPL. All labels are concatenated into one document.</summary>
        public ConversionTargetBuilder ToZpl() => To(TargetFormat.Zpl);

        /// <summary>
        /// Converts to EPL. All labels are concatenated into one document. Read
        /// <see cref="ConversionResult.Bytes"/> rather than <see cref="ConversionResult.Text"/>:
        /// EPL's <c>GW</c> command inlines raw binary.
        /// </summary>
        public ConversionTargetBuilder ToEpl() => To(TargetFormat.Epl);

        /// <summary>
        /// Converts to TSPL. All labels are concatenated into one document. As with
        /// <see cref="ToEpl"/>, prefer <see cref="ConversionResult.Bytes"/>.
        /// </summary>
        public ConversionTargetBuilder ToTspl() => To(TargetFormat.Tspl);

        /// <summary>Converts to Datamax DPL. All labels are concatenated into one document.</summary>
        public ConversionTargetBuilder ToDpl() => To(TargetFormat.Dpl);

        /// <summary>Converts to LabelZoom XML. Returns the first label only.</summary>
        public ConversionTargetBuilder ToXml() => To(TargetFormat.Xml);

        /// <summary>Converts to LabelZoom JSON. Returns the first label only; requires a paid license.</summary>
        public ConversionTargetBuilder ToJson() => To(TargetFormat.Json);

        /// <summary>Converts to PDF, one page per label.</summary>
        public ConversionTargetBuilder ToPdf() => To(TargetFormat.Pdf);

        /// <summary>Converts to a PNG image. Returns the first label only.</summary>
        public ConversionTargetBuilder ToPng() => To(TargetFormat.Png);

        /// <summary>Converts to a BMP image. Returns the first label only.</summary>
        public ConversionTargetBuilder ToBmp() => To(TargetFormat.Bmp);

        /// <summary>Converts to a GIF image. Returns the first label only.</summary>
        public ConversionTargetBuilder ToGif() => To(TargetFormat.Gif);

        /// <summary>Converts to a JPEG image. Returns the first label only.</summary>
        public ConversionTargetBuilder ToJpeg() => To(TargetFormat.Jpeg);
    }
}

File: java/src/main/java/com/labelzoom/sdk/ConversionSourceBuilder.java
package com.labelzoom.sdk;

/**
 * Chooses the target format. One class covers all thirteen.
 *
 * <p>There is no {@code toUrl}: {@code URL} is a source-only fetch instruction, and
 * {@link TargetFormat} has no constant for it. Attempting one is a compile error rather than a
 * runtime 404.
 */
public final class ConversionSourceBuilder {

    private final LabelZoomClient client;
    private final SourceFormat source;
    private final byte[] body;
    private final String contentType;

    ConversionSourceBuilder(
            LabelZoomClient client, SourceFormat source, byte[] body, String contentType) {
        this.client = client;
        this.source = source;
        this.body = body;
        this.contentType = contentType;
    }

    /** Selects the target format. */
    public ConversionTargetBuilder to(TargetFormat target) {
        return new ConversionTargetBuilder(client, source, target, body, contentType);
    }

    /** Converts to ZPL. All labels are concatenated into one document. */
    public ConversionTargetBuilder toZpl() {
        return to(TargetFormat.ZPL);
    }

    /**
     * Converts to EPL. All labels are concatenated into one document. Read
     * {@link ConversionResult#bytes()} rather than {@link ConversionResult#text()}: EPL's
     * {@code GW} command inlines raw binary.
     */
    public ConversionTargetBuilder toEpl() {
        return to(TargetFormat.EPL);
    }

    /**
     * Converts to Intermec IPL. All labels are concatenated into one document. As with
     * {@link #toEpl()}, prefer {@link ConversionResult#bytes()}: commands are framed in
     * {@code STX}/{@code ETX}.
     */
    public ConversionTargetBuilder toIpl() {
        return to(TargetFormat.IPL);
    }

    /**
     * Converts to TSPL. All labels are concatenated into one document. As with {@link #toEpl()},
     * prefer {@link ConversionResult#bytes()}.
     */
    public ConversionTargetBuilder toTspl() {
        return to(TargetFormat.TSPL);
    }

    /** Converts to Datamax DPL. All labels are concatenated into one document. */
    public ConversionTargetBuilder toDpl() {
        return to(TargetFormat.DPL);
    }

    /**
     * Converts to SATO SBPL. All labels are concatenated into one document. As with
     * {@link #toEpl()}, prefer {@link ConversionResult#bytes()}: every command is framed with
     * {@code ESC}.
     */
    public ConversionTargetBuilder toSbpl() {
        return to(TargetFormat.SBPL);
    }

    /** Converts to LabelZoom XML. Returns the first label only. */
    public ConversionTargetBuilder toXml() {
        return to(TargetFormat.XML);
    }

    /** Converts to LabelZoom JSON. First label only; requires a paid license. */
    public ConversionTargetBuilder toJson() {
        return to(TargetFormat.JSON);
    }

    /** Converts to PDF, one page per label. */
    public ConversionTargetBuilder toPdf() {
        return to(TargetFormat.PDF);
    }

    /** Converts to a PNG image. Returns the first label only. */
    public ConversionTargetBuilder toPng() {
        return to(TargetFormat.PNG);
    }

    /** Converts to a BMP image. Returns the first label only. */
    public ConversionTargetBuilder toBmp() {
        return to(TargetFormat.BMP);
    }

    /** Converts to a GIF image. Returns the first label only. */
    public ConversionTargetBuilder toGif() {
        return to(TargetFormat.GIF);
    }

    /** Converts to a JPEG image. Returns the first label only. */
    public ConversionTargetBuilder toJpeg() {
        return to(TargetFormat.JPEG);
    }
}
